### PR TITLE
Support changing pageserver dynamically

### PR DESCRIPTION
**Pull request by save-buffer**
## Problem
We currently require full restart of compute if we change the pageserver url
## Summary of changes
Makes it so that we don't have to do a full restart, but can just send SIGHUP
## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: control_plane/src/bin/neon_local.rs [hunk 1]
```diff
@@ -798,6 +798,16 @@ fn handle_endpoint(ep_match: &ArgMatches, env: &local_env::LocalEnv) -> Result<(
                 ep.start(&auth_token, safekeepers, remote_ext_config)?;
             }
         }
+        "sighup" => {
```
Comment (vadim2404) [inline control_plane/src/bin/neon_local.rs:801]: you also need to make changes in `compute_ctl` when new config arrives

File: control_plane/src/endpoint.rs [hunk 2]
```diff
@@ -621,6 +621,11 @@ impl Endpoint {
         }
     }
 
+    pub fn sighup(&self) -> Result<()> {
+        self.pg_ctl(&["reload"], &None)?;
```
Comment (vadim2404) [inline control_plane/src/endpoint.rs:625]: look at VM monitor. It executes `pg_reload_conf`. Should we follow the way?

File: control_plane/src/bin/neon_local.rs (hunk 1 above)
Comment (save-buffer) [reply control_plane/src/bin/neon_local.rs]: `compute_ctl` already has reconfigure which applies config

File: control_plane/src/endpoint.rs (hunk 2 above)
Comment (save-buffer) [reply control_plane/src/endpoint.rs]: I've changed it to pass through to `compute_ctl`.

File: control_plane/src/bin/neon_local.rs [hunk 3]
```diff
@@ -798,6 +798,24 @@ fn handle_endpoint(ep_match: &ArgMatches, env: &local_env::LocalEnv) -> Result<(
                 ep.start(&auth_token, safekeepers, remote_ext_config)?;
             }
         }
+        "reconfigure" => {
+            let endpoint_id = sub_args
+                .get_one::<String>("endpoint_id")
+                .ok_or_else(|| anyhow!("No endpoint ID provided to sighup"))?;
```
Comment (vadim2404) [inline control_plane/src/bin/neon_local.rs:804]: `to sighup` => `to reconfigure`

File: test_runner/regress/test_change_pageserver.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    row = cur.fetchone()
+    assert row is not None
+    log.info(f"shared_buffers is {row[0]}, table size {row[1]}")
+    assert int(row[0]) < int(row[1])
+
+    cur.execute("SELECT count(*) FROM foo")
+    assert cur.fetchone() == (100000,)
+
+    endpoint.reconfigure(pageserver_id=alt_pageserver_id)
+    env.pageservers[
```
Comment (vadim2404) [inline test_runner/regress/test_change_pageserver.py:51]: Check that the settings of pageservers were also changed.

File: control_plane/src/bin/neon_local.rs (hunk 3 above)
Comment (save-buffer) [reply control_plane/src/bin/neon_local.rs:804]: Done

File: test_runner/regress/test_change_pageserver.py (hunk 4 above)
Comment (save-buffer) [reply test_runner/regress/test_change_pageserver.py:59]: I assume you meant check that the GUC was changed - implemented that

Comment (vadim2404) [reply test_runner/regress/test_change_pageserver.py:59]: exactly!

File: pgxn/neon/libpagestore.c [hunk 5]
```diff
@@ -64,6 +65,20 @@ bool	(*old_redo_read_buffer_filter) (XLogReaderState *record, uint8 block_id) =
...
+
+static void
+pageserver_sighup_handler(SIGNAL_ARGS)
+{
+	if (prev_signal_handler)
+	{
+        	prev_signal_handler(postgres_signal_arg);
+	}
+	neon_log(LOG, "Received SIGHUP, flushing existing pageserver. New pageserver connstring is %s", page_server_connstring);
+	pageserver_flush();
```
Comment (knizhnik) [inline pgxn/neon/libpagestore.c:78]: Why it is necessary to flush buffered requests to PS?
We are in any case not going to receive responses, aren't we?

**Review by knizhnik — APPROVED** [review]

File: pgxn/neon/libpagestore.c (hunk 5 above)
Comment (save-buffer) [reply pgxn/neon/libpagestore.c:78]: Good point, maybe just disconnect is sufficient?

**Pull request merged**